core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/supply/AbstractVariableDescriptorBasedDemand.java

```diff
@@ -0,0 +1,49 @@
+package org.optaplanner.core.impl.domain.variable.supply;
+
+import java.util.Objects;
+
+import org.optaplanner.core.impl.domain.variable.descriptor.VariableDescriptor;
+
+/**
+ * Some {@link Demand} implementation classes are defined by their {@link VariableDescriptor} and nothing else.
+ * However, they still must not have the same {@link #hashCode()} as other {@link Demand} implementation classes
+ * defined by the same {@link VariableDescriptor}.
+ * Otherwise these different {@link Demand} implementations would be considered identical in hash maps
+ * and the same supply would be used for them.
```

Comment:
This is slightly inaccurate and may be confusing for beginners or non-Java programmers. HashCode determines the bucket. If non-equal keys have the same hashCode the lookup will be less efficient but it's not "illegal". I would rephrase this to state that different Demand impls must not be equal.